### PR TITLE
Add ST_EnvelopeAsPts spatial function

### PR DESCRIPTION
**Pull request by keyaopan**

### CODE REVIEW

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 1]
```diff
@@ -629,6 +632,26 @@ public static Slice stEnvelope(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
         return serialize(envelope);
     }
 
+    @Description("Returns the lower left and upper right corners of bounding rectangular polygon of a Geometry")
+    @ScalarFunction("ST_EnvelopeAsPts")
+    @SqlType("array(" + GEOMETRY_TYPE_NAME + ")")
+    public static Block stEnvelopeAsPts(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
+    {
+        Envelope envelope = deserializeEnvelope(input);
+        BlockBuilder blockBuilder = VARCHAR.createBlockBuilder(null, 2);
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Why VARCHAR? Shouldn't it be a GEOMETRY_TYPE?

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 2]
```diff
@@ -629,6 +632,26 @@ public static Slice stEnvelope(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+        if (envelope == null) {
+            VARCHAR.writeSlice(blockBuilder, EMPTY_MULTIPOINT);
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: `ST_EnvelopeAsPts` should return null when input geometry is empty.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 3]
```diff
@@ -629,6 +632,26 @@ public static Slice stEnvelope(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+            VARCHAR.writeSlice(blockBuilder, EMPTY_MULTIPOINT);
+        }
+        else {
+            Point point1 = new Point(envelope.getXMin(), envelope.getYMin());
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Perhaps, for clarity, rename `point1` and `point2` to `lowerLeftCorner` and `upperRightCorner`.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 4]
```diff
@@ -478,6 +480,19 @@ public void testSTEnvelope()
         assertFunction("ST_AsText(ST_Envelope(ST_GeometryFromText('GEOMETRYCOLLECTION (POINT (5 1), LINESTRING (3 4, 4 4))')))", VARCHAR, "POLYGON ((3 1, 5 1, 5 4, 3 4, 3 1))");
     }
 
+    @Test
+    public void testSTEnvelopeAsPts()
+    {
+        assertFunction("transform(ST_EnvelopeAsPts(ST_GeometryFromText('MULTIPOINT (1 2, 2 4, 3 6, 4 8)')), x -> ST_AsText(x))", new ArrayType(VARCHAR), ImmutableList.of("POINT (1 2)", "POINT (4 8)"));
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: To reduce copy-paste and improve readability, consider adding a helper method:

```
assertEnvelopeAsPts("MULTIPOINT (1 2, 2 4, 3 6, 4 8)", new Point(1, 2), new Point(4, 8));

private void assertEnvelopeAsPts(String wkt, Point lowerLeftCorner, Point upperRightCorner)
{
     assertFunction(format("transform(ST_EnvelopeAsPts(ST_GeometryFromText('%s')), x -> ST_AsText(x))", wkt), new ArrayType(VARCHAR), ImmutableList.of(new OGCPoint(lowerLeftCorner).asText(), new OGCPoint(upperRightCorner).asText()));
}
```

Please, add a test for empty geometry, e.g. `POLYGON EMPTY`.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java (hunk 1 above)
Comment (keyaopan) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Done

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java (hunk 2 above)
Comment (keyaopan) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Done

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java (hunk 3 above)
Comment (keyaopan) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Done

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java (hunk 4 above)
Comment (keyaopan) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: Done

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 5]
```diff
@@ -109,6 +109,10 @@ Operations
 
     Returns the bounding rectangular polygon of a geometry.
 
+.. function:: ST_EnvelopeAsPts(Geometry) -> Geometry
+
+    Returns the lower left and upper right corners of bounding rectangular polygon of a geometry.
```
Comment (mbasmanova) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: Let's clarify that return type is an array.

`Returns an array of two points: the lower left and upper right corners of the bounding rectangular polygon of a geometry. Returns null if input geometry is empty.`

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 6]
```diff
@@ -478,6 +482,24 @@ public void testSTEnvelope()
         assertFunction("ST_AsText(ST_Envelope(ST_GeometryFromText('GEOMETRYCOLLECTION (POINT (5 1), LINESTRING (3 4, 4 4))')))", VARCHAR, "POLYGON ((3 1, 5 1, 5 4, 3 4, 3 1))");
     }
 
+    @Test
+    public void testSTEnvelopeAsPts()
+    {
+        assertEnvelopeAsPts("MULTIPOINT (1 2, 2 4, 3 6, 4 8)", new Point(1, 2), new Point(4, 8));
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: Thanks for testing so many geometry types. For completes, let's add a test for a point: `POINT (1 2)`.

File: presto-docs/src/main/sphinx/functions/geospatial.rst (hunk 5 above)
Comment (keyaopan) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: Done

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java (hunk 6 above)
Comment (keyaopan) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: Done

**Comment by electrum** [comment]
Is this function name part of the specification? If not, we should name it “points” as we don’t abbreviate names like this (especially only saving a few characters).

**Pull request merged**